zeebe/msgpack-value/src/main/java/io/camunda/zeebe/msgpack/value/DeltaEncodedLongArrayValue.java

```diff
@@ -0,0 +1,132 @@
+ * but the difference between subsequent values is small and can be encoded with fewer bits.
+ *
+ * <p>Values are not delta encoded in memory, only when {@link
+ * DeltaEncodedLongArrayValue#write(MsgPackWriter) writing in MsgPack format} or, conversely, {@link
+ * DeltaEncodedLongArrayValue#read(MsgPackReader) reading from MsgPack format}, delta encoding is
+ * used.
+ *
+ * <p>{@link DeltaEncodedLongArrayValue#writeJSON(StringBuilder) Writing in JSON format} does not
+ * use delta encoding to keep this optimization internal and not leak to external systems. JSON is
+ * not space-efficient in any case, so we'd rather keep the delta-encoding internal as much as
+ * possible.
+ *
+ * @implNote This implementation relies on {@link MsgPackWriter#writeInteger(long)} automatically
+ *     picking the smallest possible encoding for the given deltas. By implementing this value type
+ *     in terms of a plain MsgPack array and integer values, we keep the implementation simple and
+ *     avoid the need to implement a custom encoding scheme. We lose a bit of efficiency by not
+ *     specializing for a fixed encoding of delta values but gain in flexibility because this type
+ *     can be used even in cases where delta encoding is not efficient because deltas are large
+ *     enough to require full 64-bit encoding.
+ *     <p>This encoding is a classic space-time trade-off: we save space by encoding the values as
+ *     deltas, but pay in time because reading, writing and calculating encoded size requires long
+ *     addition/subtraction and array lookbehind. We assume that this is easily optimized by any JVM
+ *     and that the space savings are worth the trade-off.
+ */
+public final class DeltaEncodedLongArrayValue extends BaseValue {
+  private long[] values;
+
+  @Override
+  public void writeJSON(final StringBuilder builder) {
+    builder.append("[");
+
+    for (var i = 0; i < values.length; i++) {
+      if (i > 0) {
+        builder.append(",");
+      }
+      builder.append(values[i]);
+    }
+
+    builder.append("]");
+  }
+
+  @Override
+  public void write(final MsgPackWriter writer) {
+    writer.writeArrayHeader(values.length);
+
+    for (int i = 0; i < values.length; i++) {
+      if (i == 0) {
+        writer.writeInteger(values[i]);
+      } else {
+        writer.writeInteger(values[i] - values[i - 1]);
+      }
+    }
+  }
+
+  @Override
+  public void read(final MsgPackReader reader) {
+    final int length = reader.readArrayHeader();
+
+    values = new long[length];
+
+    for (int i = 0; i < length; i++) {
+      if (i == 0) {
+        values[i] = reader.readInteger();
+      } else {
+        values[i] = values[i - 1] + reader.readInteger();
+      }
+    }
+  }
+
+  @Override
+  public int getEncodedLength() {
+    final int header = MsgPackWriter.getEncodedArrayHeaderLenght(values.length);
+    int data = 0;
+    for (int i = 0; i < values.length; i++) {
```

Comment:
That's practically the same that a normal `ArrayValue` does. The new add/sub shouldn't matter much and is hopefully amortized because we now work on a plain `long[]` instead of List<LongValue>`.